src/main/utility/counter.rs

```diff
@@ -24,45 +25,76 @@ pub struct Counter {
 
 impl Counter {
     /// Initializes a new counter map that starts with no keys.
-    fn new() -> Counter {
+    pub fn new() -> Counter {
         Counter {
             items: BTreeMap::new(),
         }
     }
 
     /// Increment the counter value for the key given by id.
     /// Returns the value of the counter after it was incremented.
-    fn add_one(&mut self, id: &str) -> u64 {
-        let count = self.items.entry(id.to_string()).or_insert(0);
-        *count = *count + 1;
-        *count
+    pub fn add_one(&mut self, id: &str) -> u64 {
+        match self.items.get_mut(id) {
+            Some(val) => {
+                // Increment and return the existing value without allocating new key
+                *val = *val + 1;
+                *val
+            }
+            None => {
+                // Allocate new key and insert it with initial count of 1
+                assert_eq!(self.items.insert(id.to_string(), 1), None);
+                1
+            }
+        }
     }
 
     /// Returns the counter value for the key given by id, or 0 if the key has not
     /// previously been incremented.
-    // Used for unit test
-    #[allow(dead_code)]
-    fn get_value(&mut self, id: &str) -> u64 {
-        match self.items.get(&id.to_string()) {
+    pub fn get_value(&mut self, id: &str) -> u64 {
+        match self.items.get(id) {
             Some(val) => *val,
             None => 0,
         }
     }
+}
 
-    /// Returns a string representation of the counter. The string starts with "Counts:"
-    /// and then lists known keys and values as a space-separated list of "key=value"
-    /// pairs sorted by value with the largest value first.
-    fn to_string(&mut self) -> String {
+impl Display for Counter {
+    /// Returns a string representation of the counter in the form
+    ///   `{key1:value1, key2:value2, ..., keyN:valueN}`
+    /// for known keys and values, where the list is sorted by value with the
+    /// largest value first.
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
         // Sort the counts with the heaviest hitters first
         let mut item_vec = Vec::from_iter(&self.items);
         item_vec.sort_by(|&(_, a), &(_, b)| b.cmp(&a));
 
-        // Create a string representation of the counts
-        let mut string = String::from("Counts:");
-        for item in item_vec.iter() {
-            string.push_str(format!(" {}={}", item.0, item.1).as_str());
+        // Create a string representation of the counts by iterating over the items.
+        let len = item_vec.len();
+
+        // Opening bracket.
+        match write!(f, "{{") {
+            Ok(_) => (),
+            Err(e) => return Err(e),
         }
-        string
+
+        for i in 0..len {
+            if i == (len - 1) {
```

Comment:
Would be a little nicer to dedupe by unconditionally writing the item, and then conditionally writing the `,`

Reply:
I was going to do that, but that takes 2 writes for all except the last item instead of one. Or maybe I don't have to worry about write overhead because the compiler will optimize it anyway?

Reply:
OK, done now.